Rainbow dqn

fernandotorch opened the pull request:
Barebones implementation of Rainbow dqn adapted from Kaixhin repo. Supports ai2thor and atari environments.
Fixed and extended the basic task functionality to test properly the algorithm.
Fixed config reader bug and add rendering options flag. Change defaults for performance optimizations.

fernandotorch committed:
Add Rainbow DQN code

Adapted from RL-adventure repo. First running version, although results look odd.

fernandotorch committed:
Merge branch 'A3C_barebones' into Rainbow-DQN

fernandotorch committed:
Fix channel dimension from state

There was a bug on the channel order returned from the environment by the function step. 
Rainbow dqn code also adapted to use the fix

fernandotorch committed:
Switch to Kaixhin repo

Due to the bugs found in RL-Adventure repo, the best decision was to start from a cleaner base

fernandotorch committed:
Fix to minimum working version

Working except from the CNN that picks always the same set of actions.

fernandotorch committed:
Merge branch 'master' into Rainbow-DQN

fernandotorch committed:
Fix minor bugs

Still having troubles to learn, but generally works

fernandotorch committed:
Fix to first working version

Fixed several bugs and changed the noise ratio from noisy layers in order to be able to explore the environment to learn on a basic test case.

fernandotorch committed:
Add documentation

Comments  on code + update README

fernandotorch committed:
Add numpy for general speed up

Update from last Kaixhin's pull request

fernandotorch committed:
Fix config and task

Added rendering options in config for optimization
Allow tasks to assign rewards to the different objects
Call rainbow setup from config
Ignore graphics and weights stored in the repo

fernandotorch committed:
Add support for atari games

As in the original repository. It doesn't break but it wasn't intensively tested

fernandotorch committed:
Change default params

The default params are set with the values that made ai2thor solve cup picking, the basic task

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+        self.atoms = args.atoms
+        self.Vmin = args.V_min
+        self.Vmax = args.V_max
+        # Support (range) of z
```
i barely know distributional but maybe add a few words on what "support", "range" and "z" are.
Also, add a docstring for the class?

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+"""
+Wrapper including setup, training and evaluation functions for Rainbow DQN model
```
Mention "Adapted from Kaixhin" as in other file

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+
+        self.online_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        if args.model and os.path.isfile(args.model):
+            # Always load tensors onto CPU by default, will shift to GPU if necessary
```
automatically shifts to GPU? How? why? Why is default cpu?

fernandotorch replied:
Are you happy if I change the comment to this: ```
             Always load tensors onto CPU by default, will shift to GPU if necessary to avoid 
            GPU RAM surge when loading a model checkpoint as recommended in pytorch official 
            documentation.
            Source: https://pytorch.org/docs/stable/torch.html#torch.load ```

beduffy replied:
wow this is all new to me. maybe instead of saying "if necessary" say that when you call load_state_dict it actually puts it onto the GPU at the end. It's only on the CPU at the start before fully going to GPU?

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+
+    def learn(self, mem):
+        """
+        Executes 1 gradient descent step sampling batch_size transitions form the memory
```
typo. form to from

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+            # Sample new target net noise
+            self.target_net.reset_noise()
+            # Probabilities p(s_t+n, ·; θtarget)
+            pns = self.target_net(next_states)
```
what is all this madness? can we write a book in between to explain it all? Or meet me one 10th of the way

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+            l, u = b.floor().to(torch.int64), b.ceil().to(torch.int64)
+            # Fix disappearing probability mass when l = b = u (b is int)
+            l[(u > 0) * (l == u)] -= 1
+            u[(l < (self.atoms - 1)) * (l == u)] += 1
```
i don't understand any of the above

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,134 @@
+            # m_l = m_l + p(s_t+n, a*)(u - b)
+            m.view(-1).index_add_(0, (l + offset).view(-1), (pns_a * (u.float() - b)).view(-1))
+            # m_u = m_u + p(s_t+n, a*)(b - l)
+            m.view(-1).index_add_(0, (u + offset).view(-1), (pns_a * (b - l.float())).view(-1))
```
wat

beduffy commented on algorithms/rainbow/env.py:
```diff
@@ -0,0 +1,138 @@
+        self.env = env
+        self.n_steps = n_steps
+        self.device = device
+        h, w = self.env.config['resolution'][0], self.env.config['resolution'][1]
```
is this specific to ai2thor? If yes, then it won't work on atari anymore? Are you sure you tested both atari and ai2thor together?

fernandotorch replied:
It won't work in atari but it is also not needed. Atari has it's own stuff and this is a modified lightweight version for ai2thor. I will add that to the docstrings

beduffy commented on algorithms/rainbow/env.py:
```diff
@@ -0,0 +1,138 @@
+        self.state_buffer = deque([], maxlen=n_steps)
+
+    def step(self, action):
+        # Repeat action n_step times, max pool over last 2 frames
```
I don't see how it is max pool? Where's the max, it's just taking the last and 2nd last state?

fernandotorch replied:
It means that the pool maximmum size is 2 although we move more times because history length can eventually be > 2. Will explain better

beduffy replied:
if your history length is 10 do we still only return 2 frames?
I don't understand how that is a max pool if you're just choosing specific frames?

fernandotorch replied:
Discussed on slack

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+`python main.py`
+
+Runs Rainbow DQN on our AI2ThorEnv wrapper with default params. Optionally it can be run on any
+atari environment as well using the "game" flag, e.g. --game space_invaders.
```
is it not `SpaceInvaders-v0` ?

fernandotorch replied:
As shocked as you might be, space_invaders works. Since this is a modified version from openai, they might have added the names as described in the ataripy repo https://github.com/openai/atari-py/tree/master/atari_py/atari_roms. How should we go on this one?
We can link the repo to tell people to look for more roms there, but I would certainly not list all of them

beduffy replied:
Maybe choose another example like breakout-v4 or something?

fernandotorch replied:
Chose seaquest

beduffy reviewed (commented):
Got half way, will continue in a bit.

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+parser.add_argument('--learn-start', type=int, default=int(20e3), metavar='STEPS',
+                    help='Number of steps before starting training')
+parser.add_argument('--evaluate', action='store_true', help='Evaluate only')
+parser.add_argument('--evaluation-interval', type=int, default=1e5, metavar='STEPS',
```
why are you using 1e5 and so much exponential notation everywhere?
10000 is easier to read, no?

fernandotorch replied:
It is harder to tell if there are 5 or 6 zeros right away with this order of magnitude. If you still complain after the explanation I will change it. Let me know

beduffy replied:
meh, fine. but I always found exponential harder to read than counting 0s. 1e5 means five 0s but one 1 so 6 numbers so 10k,

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+
+    # Environment selection
+    if args.game == 'ai2thor':
+        env = MultipleStepsEnv(AI2ThorEnv(
```
i know history length is 1 and you want it to work for others. But doesn't `env.env` become tiresome after a while? It might be more relevant when we go to continuous incremental rotation though.

fernandotorch replied:
Indeed. I fixed it adding a config parameter to the wrapper class so that you can access directly using env.config

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+    mem = ReplayMemory(args, args.memory_capacity)
+    priority_weight_increase = (1 - args.priority_weight) / (args.T_max - args.learn_start)
+
+    # Construct validation memory
```
2 replay memories and one for validating the loss on to test generalisation?
As well as just evaluating a trained agent on the actual environment?

fernandotorch replied:
Yes, you want to evaluate on data that wasn't seen for training. As usual in CV for training and validation splits

beduffy commented on gym_ai2thor/__init__.py:
```diff
@@ -1,4 +1,4 @@
 
-register(id='ai2thor_env-v0',
-         entry_point='gym_foo.envs:AI2ThorEnv')
+register(id='ai2thor-v0',
```
what was wrong with ai2thor_env-v0? This looks like actual ai2thor, remember we only made a wrapper

fernandotorch replied:
`mtank_ai2thor-v0` then?

beduffy replied:
cool

beduffy commented on gym_ai2thor/config_files/config_example.json:
```diff
@@ -18,6 +18,6 @@
     "task": {
         "task_name": "PickUp",
-        "target_object": "Mug"
+        "target_objects": {"Mug": 1}
```
does the 1 mean how much reward you get for that object?
Not that intuitive hmm, rewards could be a separate list maybe

fernandotorch replied:
I find it more tiresome to make 2 lists: `["Apple", "Mug", "Pencil"], [2, 1, 2]` rather than put both of them together in a dictionary. Having a dictionary forces you to write down how much reward you want to give. Explicit better than implicit

beduffy replied:
yeah but maybe we should also allow just a list which implies all objects are 1 reward?
it's a special case but also nice.

fernandotorch replied:
We can discuss this but I don't see it as a good option

beduffy replied:
For my natural language pick up or look at tasks, I'm currently not doing this but we'll see if it is necessary

fernandotorch replied:
Ok please confirm that this is the way to go. Explicit rewards and no option for list of 1 rewards.

beduffy replied:
open a thread in slack

beduffy commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -74,6 +74,11 @@ def __init__(self, seed=None, config_file='config_files/config_example.json', co
+        self.render_options = defaultdict(lambda: False)
+        if 'render_options' in self.config:
+            for option, value in self.config['render_options'].items():
+                self.render_options[option] = value
```
What are they set as default to if you don't set them?
defaultdict won't set them automatically to false until they are actually set

beduffy replied:
Wow, I was wrong, the first time you access anything e.g.
`self.render_options['fsdaf;eangeaneagea'] == False`
But maybe this isn't too clear? maybe  I'm happy to leave it

fernandotorch replied:
I set the defaultdict, builtin from python, so that if there is a KeyError you get 'False' automatically no matter which key. Standard and expected python behaviour, but let me know if you want me to document something there.

beduffy replied:
add one comment `# set segmentation and bounding box options off as default`

fernandotorch replied:
Added comment

beduffy commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -175,14 +180,16 @@ def preprocess(self, img):
         img = img.astype(np.float32)
         if self.observation_space.shape[0] == 1:
             img = rgb2gray(img)  # todo cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+        img = np.moveaxis(img, 2, 0)
```
oh shit, did we just never do this before?
Swapping RGB to BGR or vice versa. Does it make any difference?

fernandotorch replied:
Didn't try to compare, but I assume it didn't change much. I would leave it though to explicitly say that you should be careful with RGB/BGR stuff due to opencv.

beduffy replied:
ok

beduffy commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -175,14 +180,16 @@ def preprocess(self, img):
         self.event = self.controller.step(dict(action='Initialize', gridSize=0.25,
-                                               renderDepthImage=True, renderClassImage=True,
-                                               renderObjectImage=True))
+                                               renderDepthImage=self.render_options['depth'],
```
my point again above about default dict not setting these if they aren't in config

fernandotorch replied:
It is set to false automatically. Do you want me to change something or add more documentation? If so, where?

beduffy commented on gym_ai2thor/image_processing.py:
```diff
@@ -9,4 +9,4 @@ def rgb2gray(rgb):
     https://en.wikipedia.org/wiki/Luma_(video)
     """
-    return np.dot(rgb[..., :3], [0.299, 0.587, 0.114])
+    return np.expand_dims(np.dot(rgb[..., :3], [0.299, 0.587, 0.114]), axis=2)
```
hmm i have to study this. why did it work before? What are you doing?

fernandotorch replied:
You have a (h, w, 3) image and want an output of (h, w, 1) so that you can use the same code for both grayscale and RGB computations. At least that was my idea, just adding the dimension at the end so it is more "standard"

beduffy replied:
Before we were going from (h, w, 3) to (h, w)? 
If so, it might break some code elsewhere where I'm adding unsqueezes, we'll have to check

fernandotorch replied:
open a bugfix branch for that on a3c. You should get an image of (h,w,1) so that checks can be done on the 3rd channel.

beduffy replied:
ill just push do it on the new branch

beduffy commented on gym_ai2thor/tasks.py:
```diff
@@ -19,11 +19,16 @@ def create_task(config):
         if task_name == 'PickUp':
-            if config['task']['target_object'] in config['pickup_objects']:
-                return PickupTask(**config['task'])
+            # check that target objects are not selected as NON pickupables
```
yessss. I wanted to do this

beduffy commented on gym_ai2thor/tasks.py:
```diff
@@ -36,7 +41,7 @@ def __init__(self, config):
         self.max_episode_length = config['max_episode_length'] \
             if 'max_episode_length' in config else 1000
-        self.movement_reward = config['movement_reward'] if 'movement_reward' in config else 0
+        self.movement_reward = config['movement_reward'] if 'movement_reward' in config else -1
```
-1 is too big maybe. and why not default 0?

fernandotorch replied:
-1 is a common practice and helps learn faster. If you prefer I can roll back to 0 and add the movement reward to rainbow_config so that it is clear how I trained.

beduffy replied:
how about -0.1 or -0.01?
-1 is really really big. 
but sure you can add it to the config

beduffy commented on algorithms/rainbow/test.py:
```diff
@@ -0,0 +1,93 @@
+    return avg_reward, avg_Q
+
+
+def _plot_line(xs, ys_population, title, path=''):
```
have you used this?
plotly... very esoteric but im curious to see how easy it works or good it looks

fernandotorch replied:
Yes, I generated those nice graphics with mean and std of the trainings from Kaixhin's code with this. Maybe we can use another library, but this code is already there ready to be used.

beduffy commented on algorithms/rainbow/test.py:
```diff
@@ -0,0 +1,93 @@
+from algorithms.rainbow.env import Env
+
+# Globals
+Ts, rewards, Qs, best_avg_reward = [], [], [], -1e10
```
What is "Ts"?

beduffy commented on algorithms/rainbow/test.py:
```diff
@@ -0,0 +1,93 @@
+            if done:
+                state, reward_sum, done = env.reset(), 0, False
+
+            action = dqn.act_e_greedy(state)  # Choose an action ε-greedily
```
doing e-greedy or actually just greedy?
and how does this compare to training with noisy layers?

fernandotorch replied:
Will document better

fernandotorch replied:
Added this comment `# In evaluation we choose actions ε-greedily instead while fixing the noisy layers`. We could do the noisy thing, but that is what Kaixhin did and it works fine for me as well. Do you want me to change it to reset_noise to sample new weights and then feed forward the model?

beduffy replied:
Why would you evaluate e-greedily and not greedily? That's crazy, a random action has no benefit for us in evaluation, we don't need to explore in training?

As for resetting noise for EACH single step? That's a lot of exploration, not sure if that's a good idea. DQN is an off-policy algorithm so we're trying to optimise the greedy policy not the e-greedy and maybe not even the noisy layer greedy? Sooooo, maybe it's up to the user and add it to args? Or maybe it should never be done?

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        x = torch.randn(size)
+        return x.sign().mul_(x.abs().sqrt_())
+
+    def reset_noise(self):
```
I don't really understand this function and scale_noise. docstrings to explain maybe

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+
+    def forward(self, layer_input):
+        if self.training:
+            return F.linear(layer_input, self.weight_mu + self.weight_sigma * self.weight_epsilon,
```
only in training is it random?

beduffy replied:
I'm just so surprised it isn't a stochastic function.

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+            module.reset_noise()
+
+    @staticmethod
+    def get_linear_size(args):
```
dont pass args? Just pass resolution?

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        self.action_space = action_space.n
+        self.linear_in = self.get_linear_size(args)
+
+        self.conv1 = nn.Conv2d(args.history_length, 32, 8, stride=4, padding=1)
```
yeah i know history length is 1 but isn't the image rgb so 3 channels?
And if history length was 4 does that make it 12 channels or?

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        v = self.fc_z_v(F.relu(self.fc_h_v(x)))  # Value stream
+        a = self.fc_z_a(F.relu(self.fc_h_a(x)))  # Advantage stream
+        v, a = v.view(-1, 1, self.atoms), a.view(-1, self.action_space, self.atoms)
+        q = v + a - a.mean(1, keepdim=True)  # Combine streams
```
haha that one gave me nightmares for a while :P. 
Well it still does a bit :P

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        v, a = v.view(-1, 1, self.atoms), a.view(-1, self.action_space, self.atoms)
+        q = v + a - a.mean(1, keepdim=True)  # Combine streams
+        if log:  # Use log softmax for numerical stability
+          q = F.log_softmax(q, dim=2)  # Log probabilities with action over second dimension
```
do we ever use log? and why?

fernandotorch replied:
Numerical stability. I will document exactly when do we want it and why. Good point

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        if log:  # Use log softmax for numerical stability
+          q = F.log_softmax(q, dim=2)  # Log probabilities with action over second dimension
+        else:
+          q = F.softmax(q, dim=2)  # Probabilities with action over second dimension
```
why over the 2nd dimension? and why are there so many dimension in a logit? this is something to do with atoms i assume

fernandotorch replied:
Ok I can walk you through the process with docstrings, but it might be overdocumenting. The naming convention here also doesn't help, because before the softmax we don't have "q" but "z". "z" includes atoms and this softmax is taken over each Z(s,a). Will change the name and document better

beduffy commented on algorithms/rainbow/memory.py:
```diff
@@ -0,0 +1,187 @@
+        self.priority_weight = args.priority_weight
+        self.priority_exponent = args.priority_exponent
+        self.t = 0  # Internal episode timestep counter
+        # Store transitions in a wrap-around cyclic buffer within a sum tree for querying priorities
```
i suppose this almost explains why we use a segment tree, but why?
querying priorities? What are priorities? What exactly are we summing up in the tree?

beduffy commented on algorithms/rainbow/memory.py:
```diff
@@ -0,0 +1,187 @@
+import numpy as np
+
+
+# Segment tree data structure where parent node values are sum/max of children node values
```
I don't understand much of this whole file but that's ok

fernandotorch replied:
Explained more in detail. Let me know what you think

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+            T += 1
+
+            if T % args.log_interval == 0:
+                log('T = ' + str(T) + ' / ' + str(args.T_max))
```
interesting function and we might prefer to log to a file in the future so might be good to update the function but maybe not for this PR

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+                mem.priority_weight = min(mem.priority_weight + priority_weight_increase, 1)
+
+                if T % args.replay_frequency == 0:
+                    dqn.learn(mem)  # Train with n-step distributional double-Q learning
```
how does the n-step = 3 thing play into the whole DQN story?

fernandotorch replied:
TD-3 steps. That is Multi-step part of DQN. Do you need more info? I suggest you look at the paper for this first

beduffy replied:
I was trying to look in the paper. Didn't understand how we can store 3 step values into the replay buffer yet. Just seemed awkward to implement but maybe im missing something

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+
+                if T % args.evaluation_interval == 0:
+                    dqn.eval()  # Set DQN (online network) to evaluation mode
+                    avg_reward, avg_Q = test(env, T, args, dqn, val_mem)
```
again what's T?

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+                if T % args.target_update == 0:
+                    dqn.update_target_net()
+            state = next_state
+    env.close()
```
does this work in atari too?
I saw elsewhere you had a check if it was ai2thor and then did this

fernandotorch replied:
Yes this works well. In order for atari to work, we need to create an env for training and another for evaluation and that's why I close only atari env in test function. I will document that better

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+    From the paper "Noisy Networks for exploration"
+    Source: https://arxiv.org/pdf/1706.10295.pdf
+
+    This layer replaces a "linear" layer for one that describes the weights with a distribution
```
Write the noisy layer equation in here with the right symbols (unicode etc) for mu and sigma and others. Will make everything below much easier to grasp

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        self.conv1 = nn.Conv2d(args.history_length, 32, 8, stride=4, padding=1)
+        self.conv2 = nn.Conv2d(32, 64, 4, stride=2)
+        self.conv3 = nn.Conv2d(64, 64, 3)
+        self.fc_h_v = NoisyLinear(self.linear_in, args.hidden_size, std_init=args.noisy_std)
```
what is fc_h_v, fc_h_a, fc_z_v and fc_z_a? fully connected from h to value/advantage and fully connected from z to value/advantage?
What's Z? 
Can we explain this in one comment?

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        x = x.view(-1, self.linear_in)
+        v = self.fc_z_v(F.relu(self.fc_h_v(x)))  # Value stream
+        a = self.fc_z_a(F.relu(self.fc_h_a(x)))  # Advantage stream
+        v, a = v.view(-1, 1, self.atoms), a.view(-1, self.action_space, self.atoms)
```
this magic view puts everything into the right shape atom wise and we STILL do q,v a decomposition on the atoms?!

fernandotorch replied:
x = features after conv net. Then each stream tries to figure out its own atoms which will be merged into Q. What should I change here?

beduffy replied:
add a comment that we have the value and advantage DISTRIBUTIONS with the atoms reshaped here

fernandotorch replied:
view is just the equivalent of numpy.reshape for torch.Tensor variables. Changed naming for clarity, but I don't think it needs much more explanation

beduffy replied:
I know what view is?
I wanted a comment to clearly state that this isn't just Q+V decomposition, it's Q+V distribution decomposition or Q+V+atom decomposition...

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,177 @@
+    # Agent
+    dqn = Agent(args, env)
+    mem = ReplayMemory(args, args.memory_capacity)
+    priority_weight_increase = (1 - args.priority_weight) / (args.T_max - args.learn_start)
```
explain to me and maybe in a comment

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,112 @@
+        self.fc_z_v = NoisyLinear(args.hidden_size, self.atoms, std_init=args.noisy_std)
+        self.fc_z_a = NoisyLinear(args.hidden_size, self.action_space * self.atoms, std_init=args.noisy_std)
+
+    def forward(self, x, log=False):
```
forward function of rainbow is pretty simple. Looks to me like most of the complexity is in training and creation of the loss function

beduffy reviewed (commented):
Opening discussions and have actual questions in some comments

fernandotorch committed:
Add TODOs for PR

Added TODOs for each issue raised in the pull request
Minor fixes (like some comments and typos) already done

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -18,8 +20,8 @@ def __init__(self, args, action_space):
-        self.linear_in = self.get_linear_size(args)
-
+        self.linear_in = self.get_linear_size(args.resolution)
+        # TODO: check what happens if RGB. It might be 3 * history_lenght. Adapt code
```
typo lenght to length

beduffy replied:
also im really curious how frames stack on each other. 4 * history length of RGB seems strange to just stack 12 channels on top of each other

beduffy commented on README.md:
```diff
@@ -21,11 +21,16 @@ the current state of the art approaches to the problem:
+
+Implementations of these can be found in the algorithms folder and can be run on [AI2ThorEnv](https://arxiv.org/pdf/1710.02298.pdf) with:  
+
+`python algorithms/a3c/main.py`<br>
```
<br> isn't needed, you can do two spaces on the end of the line

beduffy replied:
I'm talking about the br

fernandotorch replied:
br makes it tighter, check it by yourself and let me know which one you prefer

beduffy commented on algorithms/rainbow/env.py:
```diff
@@ -0,0 +1,144 @@
+                self.frame_buffer[1] = torch.from_numpy(state)
+            if done:
+                break
+        observation = self.frame_buffer.max(0)[0]
```
Actually here is our max pool? 
im confused about this. We are actually taking the max value from each row in the 0th dimension which is of size 2 for some reason?

beduffy replied:
is this similar to a difference image? well not really but still crazy

beduffy commented on algorithms/rainbow/env.py:
```diff
@@ -0,0 +1,144 @@
+    def step(self, action):
+        """
+        Repeat action n_step times. Regardless of the number of steps, the buffer only stores
+        the last 2 frames
```
why do we move 1 step but have 2 frames? 
And we actually are doing max pool?

And again how does this work with RGB?

beduffy commented on algorithms/rainbow/env.py:
```diff
@@ -0,0 +1,144 @@
+        reward, done, info = 0, False, {}
+        for t in range(self.n_steps):
+            state, reward, done, info = self.env.step(action)
+            if t == self.n_steps - 2:
```
if n_steps was 100, we'd only take the last 2 frames? why...

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+    """
+    def __init__(self, args, env):
+        """
+        Q(s,a) is the expect reward. Z is the full distribution from which Q is generated
```
typo, expected reward

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+    def __init__(self, args, env):
+        """
+        Q(s,a) is the expect reward. Z is the full distribution from which Q is generated
+        Support represents the support of Z distribution
```
from the distributional blog:

"If p and q are two distributions with same support (i.e. their pdfs are non-zero at the same points)"
the part in brackets seems like something we can throw in

fernandotorch replied:
Isn't this going too far? explaining what the support of a random variable is. Please confirm if you really want to do this

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+        """
+        Q(s,a) is the expect reward. Z is the full distribution from which Q is generated
+        Support represents the support of Z distribution
+        Z is represented with a fixed number of "atoms", which are discrete positions equidistant
```
I don't think the atoms are the discrete positions themselves since the blog says "the positions of the atoms"
They're more like the discrete pairs of position and probability (x1,p1),…,(xN,pN)

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,178 @@
+                    help='Network hidden size')
+parser.add_argument('--noisy-std', type=float, default=0.8, metavar='σ',
+                    help='Initial standard deviation of noisy linear layers')
+parser.add_argument('--atoms', type=int, default=51, metavar='C',
```
maybe num-atoms would be better?

fernandotorch replied:
having to write num-atoms is 4 more chars on the terminal/cmd and for me it seems to clarify nothing, since you don't have more than 1 param related to atoms. But if you still want to go with this confirm

beduffy replied:
no one actually writes any parameters in haha. No one will ever change it from 51 and if they do it'll be the same length as `--noisy-std` above. 0 problems. 
People might actually get confused that these aren't the atoms, it's just the number of atoms.
Just because there's only 1 thing mentioning atoms doesn't mean it's clear that atoms == num_atoms...

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,117 @@
+        self.register_buffer('weight_epsilon', torch.empty(out_features, in_features))
+        self.bias_mu = nn.Parameter(torch.empty(out_features))
+        self.bias_sigma = nn.Parameter(torch.empty(out_features))
+        self.register_buffer('bias_epsilon', torch.empty(out_features))
```
Can you put the parameters together in their own 4 line block (with an empty line between) and then call the register buffer stuff. It's a bit hard and scary to read.

or wait. register_buffer creates params. Why do we do this?
At least put a comment before all 6 saying:
the 6 parameters bla bla

with space above and below

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+
+            # Compute Tz (Bellman operator T applied to z)
+            # Tz = R^n + (γ^n)z (accounting for terminal states)
+            Tz = returns.unsqueeze(1) + nonterminals * (self.discount ** self.n) \
```
help me understand self.n and n-step q-learning

fernandotorch replied:
You just found a bug sir, but I will answer your question regardless before I fix it.
In N-step Q-Learning (from Sutton) we make N steps and collect all those returns, then use Q to estimate the expected return, having Q_target = R1 + R2 + R3 + Q(s,a), analogous to TD n-step. Now you probably see the formula much simpler

beduffy replied:
I've always generally known this but none of the code seems to suggest that we actually store R1 + R2 + R3. Or is it in the returns itself somehow? I'm trying to do a ctrl+f on self.n... self.n is maybe a bad name too but not too bad.
I'm still confused a bit about it being called nonterminals as well. Maybe we'll take this in slack

fernandotorch replied:
The key to understanding that is mem.sample(self.batch_size) in that script. The returns output from the mem.sample function are already summed for "truncated n-steps", i.e. if multi_step parameter is set to 3, 3 consecutive steps from the memory are retrieved and their rewards summed before appending that value to returns. You do that operation until you obtain a tensor with size self.batch_size. That's why you see here only the last operation. Non-terminals is a boolean that checks if you did get enough transitions or actually got to terminal state before or right at the end (after a terminal state there is no step in theory). Hope this is clearer now

beduffy replied:
It is clearer after I looked at the _get_sample_from_segment() function from memory.py. 
You should maybe say that in a comment (don't make it two lines).
`# look in _get_sample_from_segment() from memory.py for more details`

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+    def act(self, state):
+        """ Acts based on single state (no batch) """
+        with torch.no_grad():
+            return (self.online_net(state.unsqueeze(0)) * self.support).sum(2).argmax(1).item()
```
can you go through this with me?
online net returns z or q? `q = F.softmax(q, dim=2)` is what it returns. what shape is this? put a comment in agent.py on what shape this is and say z if it is z.

ok now, why do we multiply by self.support? and why do we `.sum(2)`?

fernandotorch replied:
We covered all this in the learn function of agent.py. Please read that again and we can talk if there is any point to clarify. Should I make a comment saying `for further information go to learn function in agent.py`? 
Z(s,a) is the distribution of expected returns, Q is the mean of Z or simply expected return for each action. Z has size s * a * atoms and Q has size s * a.

beduffy replied:
hmm. I actually had to go to model.py to look further.
So online_net returns Q and nothing here involves Z then. Why are you mentioning Z to me?
And again, why do we multiply by self.support? and why do we `.sum(2)`?

fernandotorch replied:
Ok let's go slower. Z(s,a) is a distribution defined by the atoms ((x_i, p_i), ...., (x_n, p_n)). Since Q is the expectation of Z, i.e. its "mean", by definition you sum the products x_i * p_i to obtain it. Since Z has shape (1, a, num_atoms), multiplying by the x_is (support) with p_is (values) and summing the second axis gives you the Q for every action in the input state. Argmax is the greedy part. Let me know if you still have problems with this.

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,158 @@
+          mem.sample(self.batch_size)
+
+        # Calculate current state probabilities (online network noise already sampled)
+        log_ps = self.online_net(states, log=True)  # Log probabilities log p(s_t, ·; θonline)
```
log_ps? ps? Are you not returning Q?

From the blog, they mention the net returning |A| x num_atoms matrix.

fernandotorch replied:
I added this comment on model.py `log softmax used while learning to generate probabilities with higher numerical stability`. Let me know if that is enough for you.

beduffy replied:
why do we need higher numerical stability in training only? Because training can get crazy and blow up things?
Why would you ever choose the non-log softmax version?

fernandotorch replied:
This is supporting the moving target DQN. So you are chasing an optimum that is moving and you want to add stability. During evaluation you don't care anymore because that is not going to be "propagated knowledge". Ideally I think you don't want the log, but it helps like an extra learning rate (kind of, please take it as an informal statement and not taken from a calculus book).

beduffy replied:
Ok I accept your informality. 

maybe another reason is that we only need logs for backprop and many other mathematical formulations, not really for just choosing the argmax. Maybe add that in as well? e.g. `log is not needed for evaluation `

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,178 @@
+            action = dqn.act(state)  # Choose an action greedily (with noisy weights)
+            next_state, reward, done, _ = env.step(action)  # Step
+            if args.reward_clip > 0:
+                reward = max(min(reward, args.reward_clip), -args.reward_clip)  # Clip rewards
```
for environments where we have different rewards for different objects e.g. 1, 2, 5 we need to remember to not clip them

fernandotorch replied:
if clip == 0 nothing is clipped. In main.py the --help for that argument explains it. Should we make it even more explicit? Give me ideas or close this

beduffy replied:
set the default to 5 or 10?

fernandotorch replied:
I will set default to 0 better, so you explicitily have to put it if you want it. Change my mind

beduffy replied:
obviously r == 0 will fail the r > 0 but it's tight i.e. adding 0.0000001 will make it happen haha. As long as we say in the argument's help clearly, i'm fine with it. You can resolve if so

beduffy commented on algorithms/rainbow/main.py:
```diff
@@ -0,0 +1,178 @@
+            if args.reward_clip > 0:
+                reward = max(min(reward, args.reward_clip), -args.reward_clip)  # Clip rewards
+            mem.append(state, action, reward, done)  # Append transition to memory
+            T += 1
```
no way, this T is sent into test() function which has Ts and Ts_rewards. Bad variable name. Should be num_steps

fernandotorch replied:
Also changed the T from above to mem_steps, since that loop is used for a first collection of transitions for the replay memory

fernandotorch committed:
Add explanations on policy evaluation

First part explaining the calculations from distributional RL

fernandotorch committed:
Fix Multistep env wrapper

Deleted all the madness needed for atari environments and refactored to add the possibility of stacking frames which might be useful to ai2thor.

fernandotorch committed:
Finish documenting distributional DQN

Added complete description of all steps taken for multi step distributional DQN

beduffy commented on algorithms/rainbow/model.py:
```diff
@@ -0,0 +1,117 @@
+            out_conv2 = ((out_conv1 - 4) // 2) + 1
+            out_conv3 = (out_conv2 - 3) + 1
+            linear_size *= out_conv3
+        return linear_size
```
could you explain this a bit. Add more to the docstring. explain why `*=`

I dont fully get it even though I assume you got this idea from me. 
I just had to make a new one of these for A3C_LSTM_GA

I was calling it width and then at the end to calculate linear size I did
width * width * num_filters

fernandotorch committed:
Add support for RGB

Ai2thor can use now rgb or grayscale images. Atari still uses only grayscale as from original code and corresponding papers

fernandotorch committed:
Add comments for ai2thor specs

Specified in the PR

fernandotorch committed:
Add comments to noisy layers

To explain how they work in the code. Still incomplete

beduffy commented on algorithms/rainbow/test.py:
```diff
@@ -12,8 +12,11 @@
-# Globals
-# TODO: explain what is Ts and Qs
+""" Global variables used to track the evaluation results
+Ts               - list of evaluation steps at each evaluation period
```
I still think T is bad name (mentioned in other comment) here due to transformation T in the distributional algorithm and generally just being a bad name.

fernandotorch replied:
Changed it to steps

beduffy commented on gym_ai2thor/config_files/rainbow_example.json:
```diff
@@ -17,6 +17,7 @@
         "task_name": "PickUp",
-        "target_objects": {"Mug": 10}
+        "target_objects": {"Mug": 10},
+        "movement_reward": -1
```
again, movement_reward == -1 is crazy. Why not -0.1 or -0.01
You need to pick up 500 cups (optimal max since pickup and putdown) to get an optimal score of -500?

but I see in the tasks file the default is -0.01 at least. If it works for rainbow, you can resolve

fernandotorch committed:
Add more comments on distributional RL

Explained the model more clearly and cleaned up a bit more agent and env

beduffy commented on algorithms/rainbow/agent.py:
```diff
@@ -0,0 +1,179 @@
+            Probabilities: p(s_t+n, ·; θonline), i.e. for all actions
+            """
+            pns = self.online_net(next_states)
+            """ We compute the expected Q from the N-step distribution 
```
Change 2 line comments to hashtags, makes it strange to read. 
And possibly even with 3 line comments. 
Feel free to argue

fernandotorch replied:
2 line comments fine, 3 line comments no. I like more """ for long comments and it is also PEP8 so let's put that middleground

beduffy replied:
ok for now. As long as an algorithm doesn't get broken up too much by multi-line comments i.e. over documenting

fernandotorch committed:
Renaming + more comments

To clarify even more the code flow

fernandotorch committed:
Fix multi-step bug 

Not really in use, but it should work anyways

fernandotorch committed:
Fix config_dict input + more comments

Added explanations and a new argument to allow for a command line argument to change the environment details

fernandotorch committed:
Add more comments

Explain more about memory, PER and the special case of ai2thor for rendering

fernandotorch committed:
Fix arg.config_dict

Dicitonaries can't be arguments for argparse so the best solution is to input a json string and check it within ai2htor. This is an ai2thor bug and not a Rainbow error

fernandotorch committed:
Fix config json input not dict

As discussed, we allow the input of a path to a config file to define ai2thor and task parameters for rainbow using that method

fernandotorch committed:
Rename env wrapper + fix test

For clarity the wrapper name was changed
1 test corrected for the new way of declaring target objects and rewards explicitly

beduffy commented:
Instead of commenting on the code, I'll just edit this final comment as we get the final things done and we can do it together. We can also merge well before some of these are resolved as they are just questions. 

1. env.py - FrameStackEnv step better docstring
2. env.py - FrameStackEnv what if we finish midway?
3. env.py - FrameStackEnv why do you step after reset and env.reset returns state but you step?
4. model.py - why the name z_q? is it not just z?
5. model.py - it's important to say square image for get_linear_size(), otherwise it won't work right? I also prefer to get rid of the `*=` 
6. main.py extra space after first docstring and before `parser = `
~~7. main.py - you aren't saving model? Oh wait you are in test. dqn.save(). I did similar with A3C_GA using test results so I'm happy about this ~~
8. main.py - What does dqn.train()/eval() affect? usually dropout and batchnorm but here?
9. main.py - val_mem, we store empty reward and action? And we keep it the same for the entirety of training? How do we get accurate results?
10. main.py - priority_weight_increase_amt maybe. I was confused at first. 
11. test.py - two instantiations are not needed. So you're taking the training environment and suddenly changing it to a test environment? Hmmmm. Could cause problems. 
12. test.py - I'd like you to say a bit more clearly that we are testing both in environment and the same validation memory we saved at the very start of main.py. `# Test Q-values over validation memory` -> space before this line and `# And now we actually test Q-values over validation memory which was saved at the start of training`
13. test.py - not explaining T. num_steps would do.
14. test.py - `if not evaluate` do stuff if not evaluate? opposite to what I expected? 
15. test.py - what's all this trace stuff and ys stuff?
16. agent.py - Talk about higgsfields var names and projection function. Generally could make learn() a bit easier to read. Other things we could find there? Also perfect opportunity to discuss how there's a line of how jumping between files/classes/functions get's harder than the spaghetti notebook haha.


Will keep updating

fernandotorch committed:
Add todos for env and main.py

After our discussion on the issues in the PR related to those files

fernandotorch committed:
Fix buffer issues in FrameStack wrapper

Now the state buffer is filled on env.reset() and step only appends 1 frame to the end of the deque

fernandotorch committed:
Fix rgb training + add comments

Transitions weren't stored properly for RGB images. Now it works for both rgb and grayscale.
Added docstrings to explain in more detail the functions in memory

fernandotorch committed:
Fix Distributional explanations + add todos

Explained in more detail the code from the agent.py function learn

fernandotorch committed:
Abstract distributional projections

Simplify learn function by abstracting all the messiness to get the target distribution

fernandotorch committed:
Fix plotly import + comments

Added all comments in response to the issues discussed. Plotly import is now optional but required to make and save plots.

fernandotorch committed:
Merge branch 'master' into Rainbow-DQN

fernandotorch committed:
Fix comments

Indentations, typos and other explanations

fernandotorch committed:
Fix bug in evaluation

since we changed FrameStack wrapper I forgot to remove the "unsqueeze" operation.

fernandotorch committed:
Add more comments

To explain more about the distributional part of rainbow

fernandotorch committed:
Fix final comments
